examples/credential-registry-storage-contract/src/lib.rs

```diff
@@ -0,0 +1,418 @@
+    Ok(State::empty(state_builder))
+}
+
+/// View the registered credential data.
+/// It rejects if:
+/// - It fails to parse the parameter.
+#[receive(
+    contract = "credential-registry-storage",
+    name = "view",
+    parameter = "PublicKeyEd25519",
+    return_value = "SimpleCredentialState",
+    error = "CustomContractError"
+)]
+fn view<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &impl HasHost<State<S>, StateApiType = S>,
+) -> ContractResult<Option<SimpleCredentialState>> {
+    let public_key: PublicKeyEd25519 = ctx.parameter_cursor().get()?;
+
+    Ok(host.state().credential_registry.get(&public_key).map(|credential_state| {
+        SimpleCredentialState {
+            metadata:             credential_state.metadata,
+            encrypted_credential: credential_state.encrypted_credential.clone(),
+        }
+    }))
+}
+
+/// The parameter type for the contract function `store`.
+#[derive(Serialize, SchemaType, Debug)]
+pub struct StoreParam {
+    /// The contract_address that the signature is intended for.
+    contract_address:     ContractAddress,
+    /// The serialized encrypted_credential.
+    #[concordium(size_length = 2)]
+    encrypted_credential: Vec<u8>,
+    /// Metadata associated with the credential.
+    metadata:             Metadata,
+    /// Public key that created the above signature.
+    public_key:           PublicKeyEd25519,
+    /// Signature.
+    signature:            SignatureEd25519,
+    /// A timestamp to make signatures expire.
+    timestamp:            Timestamp,
+}
+
+impl StoreParam {
+    /// Prepare the message bytes for signature verification
+    fn message_bytes(&self, bytes: &mut Vec<u8>) -> ContractResult<()> {
+        self.contract_address.serial(bytes).map_err(|_| CustomContractError::SerializationError)?;
+        self.encrypted_credential
+            .serial::<Vec<_>>(bytes)
+            .map_err(|_| CustomContractError::SerializationError)?;
+        self.metadata.serial(bytes).map_err(|_| CustomContractError::SerializationError)?;
+        self.timestamp.serial(bytes).map_err(|_| CustomContractError::SerializationError)?;
+        Ok(())
+    }
+}
+
+/// The parameter type for the contract function `serializationHelper`.
+#[derive(Serialize, SchemaType)]
+pub struct SerializationHelperParam {
+    /// The contract_address that the signature is intended for.
+    contract_address:     ContractAddress,
+    /// The serialized encrypted_credential.
+    #[concordium(size_length = 2)]
+    encrypted_credential: Vec<u8>,
+    /// Metadata associated with the credential.
+    metadata:             Metadata,
+    /// A timestamp to make signatures expire.
+    timestamp:            Timestamp,
+}
+
+/// Helper function that can be invoked at the front-end to serialize
+/// the message to be signed by the wallet.
+#[receive(
+    contract = "credential-registry-storage",
+    name = "serializationHelper",
+    parameter = "SerializationHelperParam"
+)]
```

Comment:
Could you please explain a bit more about how it works on the front-end side? (If you think it makes sense to explain here)
I see that the input parameter is serialized when the entrypoint is called, but it doesn't return anything.

Reply:
In the 'old' signing mechanism, we needed to serialize the `SerializationHelperParam` (which includes all the input parameters from `StoreParam` except for the `signature` and the `public_key`). Meaning `SerializationHelperParam` includes everything the wallet is supposed to sign. In the below code line, you can see we only need the schema of the `serializationHelper` function. The `serializationHelper` function is not executed at any point in time, therefore the logic of the function is irrelevant which is why the function has no logic and just returns `Ok(())`.

https://github.com/Concordium/concordium-dapp-examples/blob/main/sponsoredTransactions/frontend/src/SponsoredTransactions.tsx#L180

The 'new' signing mechanism for the verifiable credentials is still not yet implemented and subject to changes in the wallets. I added this statement above the function but I think we will revisit this function once we have a working frontend/browser wallet example:
```
Helper function that can be invoked at the front end to serialize the `SerializationHelperParam` to be signed by the wallet. `SerializationHelperParam` includes all the input parameters from `StoreParam` except for the `signature` and the `public_key`. We need only the input parameter schema of this function at the front end. The `serializationHelper` function is not executed at any point in time, therefore the logic of the function is irrelevant.
```